File: codegen/objc/objc-entities/src/main/groovy/com/stanfy/helium/handler/codegen/objectivec/entity/builder/DefaultObjCProjectBuilder.java
package com.stanfy.helium.handler.codegen.objectivec.entity.builder;

import com.stanfy.helium.handler.codegen.objectivec.entity.ObjCHeaderFile;
import com.stanfy.helium.handler.codegen.objectivec.entity.ObjCImplementationFile;
import com.stanfy.helium.handler.codegen.objectivec.entity.ObjCProject;
import com.stanfy.helium.handler.codegen.objectivec.entity.ObjcEntitiesOptions;
import com.stanfy.helium.handler.codegen.objectivec.entity.file.ObjCClass;
import com.stanfy.helium.handler.codegen.objectivec.entity.file.ObjCClassDefinition;
import com.stanfy.helium.handler.codegen.objectivec.entity.file.ObjCClassImplementation;
import com.stanfy.helium.handler.codegen.objectivec.entity.file.ObjCPropertyDefinition;
import com.stanfy.helium.model.Field;
import com.stanfy.helium.model.Message;
import com.stanfy.helium.model.Project;
import com.stanfy.helium.model.Type;

import java.util.HashSet;
import java.util.Map;

/**
 * Created by ptaykalo on 8/17/14.
 *
 */
public class DefaultObjCProjectBuilder implements ObjCProjectBuilder {

  /*
     Type transformer to transform correct Objc types from Helium API
     */
  private ObjCTypeTransformer typeTransformer = new ObjCTypeTransformer();
  private ObjCPropertyNameTransformer nameTransformer = new ObjCPropertyNameTransformer();

  @Override
  public ObjCTypeTransformer getTypeTransformer() {
    return typeTransformer;
  }

  @Override
  public ObjCPropertyNameTransformer getNameTransformer() {
    return nameTransformer;
  }

  /*
  Performs parsing / translation of Helium DSL Proejct Structure to Objective-C Project structure
   */
  @Override
  public ObjCProject build(final Project project) {
    return build(project, null);
  }

  /*
  Performs parsing / translation of Helium DSL Proejct Structure to Objective-C Project structure
  Uses specified options for the generation @see ObjCProjectParserOptions
   */
  @Override
  public ObjCProject build(final Project project, final ObjcEntitiesOptions options) {
    ObjCProject objCProject = new ObjCProject();

    // Register all messages first
    for (Message message : project.getMessages()) {
      if (message.isAnonymous() || (options != null && !options.isTypeIncluded(message))) {
        continue;
      }
      String messageName = message.getName();
      String className = messageName;
      if (options != null && options.getPrefix() != null) {
        className = options.getPrefix() + className;
      }
      typeTransformer.registerRefTypeTransformation(messageName, className);
    }

    // Registering all custom mappings
    if (options != null && options.getCustomTypesMappings() != null) {
      Map<String, String> customTypesMappings = options.getCustomTypesMappings();
      for (Map.Entry<String, String> entry : customTypesMappings.entrySet()) {
        // Check if Objective-C Type have * here
        String objectiveCType = entry.getValue();
        String heliumType = entry.getKey();
        if (objectiveCType.contains("*")) {
          String validObjectiveCString = objectiveCType.replace("*", "").trim();
          typeTransformer.registerRefTypeTransformation(heliumType, validObjectiveCString);
        } else {
          typeTransformer.registerSimpleTransformation(heliumType, objectiveCType);
        }
      }
    }

    for (Message message : project.getMessages()) {

      String fileName = message.getName();
      if (options != null && options.getPrefix() != null) {
        fileName = options.getPrefix() + fileName;
      }

      ObjCClass objCClass = new ObjCClass(fileName);

      ObjCClassDefinition classDefinition = new ObjCClassDefinition(fileName);
      ObjCClassImplementation classImplementation = new ObjCClassImplementation(fileName);

      HashSet<String> usedPropertyNames = new HashSet<String>();
      for (Field field : message.getActiveFields()) {
        String propertyName = nameTransformer.propertyNameFrom(field.getName(), usedPropertyNames);
        Type heliumAPIType = field.getType();
        String propertyType = typeTransformer.objCType(heliumAPIType, field.isSequence());

        if (heliumAPIType instanceof Message && !field.isSequence()) {
          classDefinition.addExternalClassDeclaration(propertyType.replaceAll("\\*|\\s", ""));
        }
        ObjCPropertyDefinition.AccessModifier accessModifier = typeTransformer.accessorModifierForType(heliumAPIType);
        ObjCPropertyDefinition property = new ObjCPropertyDefinition(propertyName, propertyType, accessModifier);
        property.setCorrespondingField(field);

        if (field.isSequence()) {
          property.setComment(" sequence of " + typeTransformer.objCType(heliumAPIType, false) + " items");
          property.setSequence(true);
          property.setSequenceType(typeTransformer.objCType(heliumAPIType, false).replaceAll("\\*|\\s", ""));
        }

        classDefinition.addPropertyDefinition(property);

        // Update used Names
        usedPropertyNames.add(propertyName);
      }

      objCClass.setDefinition(classDefinition);
      objCClass.setImplementation(classImplementation);

      objCProject.addClass(objCClass, message.getName());
      ObjCHeaderFile headerFile = new ObjCHeaderFile(fileName);
      ObjCImplementationFile implementationFile = new ObjCImplementationFile(fileName);
      headerFile.addSourcePart(classDefinition);
      implementationFile.addSourcePart(classImplementation);

      objCProject.addFile(headerFile);
      objCProject.addFile(implementationFile);
    }

    // classes
    return objCProject;
  }

}

File: codegen/objc/objc-entities/src/main/kotlin/com/stanfy/helium/handler/codegen/objectivec/entity/builder/DefaultObjCProjectBuilder.kt
package com.stanfy.helium.handler.codegen.objectivec.entity.builder;

import com.stanfy.helium.handler.codegen.objectivec.entity.ObjCEntitiesOptions
import com.stanfy.helium.handler.codegen.objectivec.entity.ObjCHeaderFile
import com.stanfy.helium.handler.codegen.objectivec.entity.ObjCImplementationFile
import com.stanfy.helium.handler.codegen.objectivec.entity.ObjCProject
import com.stanfy.helium.handler.codegen.objectivec.entity.file.ObjCClass
import com.stanfy.helium.handler.codegen.objectivec.entity.file.ObjCClassInterface
import com.stanfy.helium.handler.codegen.objectivec.entity.file.ObjCImplementationFileSourcePart
import com.stanfy.helium.handler.codegen.objectivec.entity.file.ObjCPropertyDefinition
import com.stanfy.helium.model.Message
import com.stanfy.helium.model.Project

/**
 * Created by ptaykalo on 8/17/14.
 */
public class DefaultObjCProjectBuilder : ObjCProjectBuilder {

  private val _nameTransformer = ObjCPropertyNameTransformer()
  override fun getNameTransformer(): ObjCPropertyNameTransformer {
    return _nameTransformer;
  }

  private val _typeTransformer = ObjCTypeTransformer()
  override fun getTypeTransformer(): ObjCTypeTransformer {
    return _typeTransformer
  }

  /**
   * Performs parsing / translation of Helium DSL Project Structure to Objective-C Project structure
   */
  override fun build(project: Project): ObjCProject {
    return build(project, null)
  }

  /**
   * Performs parsing / translation of Helium DSL Project Structure to Objective-C Project structure
   * Uses specified options for the generation @see ObjCProjectParserOptions
   */
  override fun build(project: Project, options: ObjCEntitiesOptions?): ObjCProject {
    val objCProject = ObjCProject()

    project.messages
        .filter { message ->
          !message.anonymous && (options == null || options.isTypeIncluded(message))
        }
        .forEach { message ->
          val messageName = message.name
          val className = (options?.prefix ?: "") + messageName
          _typeTransformer.registerRefTypeTransformation(messageName, className)

          // check for custom mappings
          val customTypeMappings = options?.customTypesMappings?.entries
          if (customTypeMappings != null) {
            for ((heliumType, objcType) in customTypeMappings) {
              if (objcType.contains("*")) {
                val validObjectiveCString = objcType.replace("*", "").trim()
                _typeTransformer.registerRefTypeTransformation(heliumType, validObjectiveCString);
              } else {
                _typeTransformer.registerSimpleTransformation(heliumType, objcType);
              }
            }
          }
        }

    project.messages
        .forEach { message ->
          val filename = (options?.prefix ?: "") + message.name
          val objCClass = ObjCClass(filename)

          val classDefinition = ObjCClassInterface(filename)
          val classImplementation = ObjCImplementationFileSourcePart(filename)

          var usedPropertyNames = hashSetOf<String>()
          message.activeFields
              .map { field ->
                val propertyName = _nameTransformer.propertyNameFrom(field.name, usedPropertyNames)
                val heliumAPIType = field.type
                val propertyType = _typeTransformer.objCType(heliumAPIType, field.isSequence)
                if (heliumAPIType is Message && !field.isSequence) {
                  classDefinition.addExternalClassDeclaration(propertyType.replace("*", "").replace(" ",""));
                }

                val accessModifier = _typeTransformer.accessorModifierForType(heliumAPIType)
                val property = ObjCPropertyDefinition(propertyName, propertyType, accessModifier)
                property.correspondingField = field

                if (field.isSequence) {
                  property.comment = " sequence of " + _typeTransformer.objCType(heliumAPIType, false) + " items"
                  property.isSequence = true
                  property.sequenceType = _typeTransformer.objCType(heliumAPIType, false).replace("*", "").replace(" ","")
                }
                // Update used Names
                usedPropertyNames.add(propertyName)

                property

              }
              .forEach { property ->
                classDefinition.addPropertyDefinition(property);
              }

          objCClass.definition = classDefinition;
          objCClass.implementation = classImplementation;

          objCProject.addClass(objCClass, message.name);
          val headerFile = ObjCHeaderFile(filename);
          val implementationFile = ObjCImplementationFile(filename);
          headerFile.addSourcePart(classDefinition);
          implementationFile.addSourcePart(classImplementation);

          objCProject.addFile(headerFile);
          objCProject.addFile(implementationFile);
        }

    return objCProject
  }
}
